doc/Pacemaker_Explained/en-US/Ch-Rules.txt

```diff
@@ -522,6 +522,121 @@ You may wish to write +end="2005-03-31T23:59:59"+ to avoid confusion.
 -------
 =====
 
+== Resource Expressions ==
+
+An +rsc_expression+ is a rule condition based on a resource agent's properties.
+This rule is only valid within an +rsc_defaults+ or +op_defaults+ context.
+
+.Attributes of an rsc_expression Element
+[width="95%",cols="2m,<5",options="header",align="center"]
+|=========================================================
+
+|Field
+|Description
+
+|id
+|A unique name for the expression (required)
+ indexterm:[XML attribute,id attribute,rsc_expression element]
+ indexterm:[XML element,rsc_expression element,id attribute]
+
+|class
+|The standard name to be matched against resource agents (required)
+ indexterm:[XML attribute,class attribute,rsc_expression element]
+ indexterm:[XML element,rsc_expression element,class attribute]
+
+|provider
+|If given, the vendor to be matched against resource agents.  This
+ only makes sense for agents using the OCF spec.
+ indexterm:[XML attribute,provider attribute,rsc_expression element]
+ indexterm:[XML element,rsc_expression element,provider attribute]
+
+|type
+|The name of the resource agent to be matched (required)
+ indexterm:[XML attribute,type attribute,rsc_expression element]
+ indexterm:[XML element,rsc_expression element,type attribute]
```

Comment:
Are class and type really required? It might be reasonable for someone to set a default timeout for all stonith resources, for example, without specifying a type.

Reply:
At the moment they are both required, yes.  You can see this in pe__eval_op_expr.  It would be very easy to change, though.

Reply:
Let's do that then. I think that will be the final touch

Reply:
Okay, just finishing up some testing on this.  The changes will be throughout the patches - XML schema change, docs change, additional test case, changes to the rules code.  I didn't want to make it too easy to review.

Reply:
And pushed.